### PR TITLE
treewide: fix cffi-related cross-compilation

### PR DESCRIPTION
**Pull request by legendofmiracles**
###### Motivation for this change
I made a list of all packages listing `cffi` incorrectly (as described here: https://github.com/NixOS/nixpkgs/pull/135447) and went through all of them manually to fix it. It worked for maybe 1/10th of the packages, while I was working I removed the ones from the list that built and added comments to the ones that didn't:
```
All tests were run with nix build nixpkgs#pkgsCross.raspberryPi.*

~~Sometimes packages will cross compile, even though when I try to cross compile one of their dependencies it fails... No idea what's going on there - probably some python weirdness~~


# rust boostrap, don't want to lol
pkgs/tools/security/sequoia/default.nix
# cc command not found
pkgs/development/python-modules/angr/default.nix
# the cffi error, after fixing that: wrong ELF class: ELFCLASS32 when loading a library
pkgs/development/python-modules/augeas/default.nix
# issue doesn't go away
pkgs/development/python-modules/autobahn/default.nix
# wrong ELF class: ELFCLASS32
pkgs/development/python-modules/bedup/default.nix
# cc not found
pkgs/development/python-modules/cle/default.nix
# idk what the derivation name is
pkgs/development/python-modules/cld2-cffi/default.nix
# doesn't work
pkgs/development/python-modules/deltachat/default.nix
# doesn't work
pkgs/development/python-modules/etesync/default.nix
# /nix/store/mij848h2x5wiqkwhg027byvmf9x3gx7y-glibc-2.33-50/lib/librt.so: file not recognized: file format not recognized
pkgs/development/python-modules/google-crc32c/default.nix
# unsupported
pkgs/development/python-modules/jupyter_server/default.nix
# cffi issue is fixed, but other issue: checking for a version of Python >= '2.1.0'... ./configure: line 12491: /nix/store/.../bin/python: cannot execute binary file: Exec format
pkgs/development/python-modules/keyrings-cryptfile/default.nix
# unsupported
pkgs/development/python-modules/notebook/default.nix
# Please install the program "msgfmt" in advance.
pkgs/development/python-modules/notmuch/2.nix
# /nix/store/9qsrf2ddrd1s3myyk89w94wcyw0w3idq-postgresql-13.4-lib/lib/libpq.so: file not recognized: file format not recognized
pkgs/development/python-modules/psycopg2/default.nix
# doesn't work
pkgs/development/python-modules/pyspice/default.nix
# unsupported
pkgs/development/python-modules/pyspotify/default.nix
# not even gonna try
pkgs/development/python-modules/pytorch/default.nix
# cc no such file or directory
pkgs/development/python-modules/pyvex/default.nix
# idk billion deps and some pkg-config error
pkgs/development/python-modules/pyvips/default.nix
# can't find ar
pkgs/development/python-modules/rtmixer/default.nix
# unsupported
pkgs/development/python-modules/rpy2/default.nix
# ERROR: Could not find a version that satisfies the requirement pytest-runner (from versions: none)
pkgs/development/python-modules/weasyprint/default.nix
# stdenv-linux/setup: /nix/store/h6g8yf6p2rc36hcvshkdh23xxpc5c4kv-python3-armv6l-unknown-linux-gnueabihf-3.9.6/bin/python3.9: cannot execute binary file: Exec format error
pkgs/development/python-modules/xattr/default.nix
# stdenv-linux/setup: /nix/store/h6g8yf6p2rc36hcvshkdh23xxpc5c4kv-python3-armv6l-unknown-linux-gnueabihf-3.9.6/bin/python3.9: cannot execute binary file: Exec format error
pkgs/development/python-modules/pymunk/default.nix
# ./autogen.sh: line 29: libtoolize: not found
pkgs/development/python-modules/ocrmypdf/default.nix
# configure: error: in `/build/source':
# configure: error: cannot run test program while cross compiling
pkgs/development/python-modules/kaldi-active-grammar/default.nix
# wrong ELF class: ELFCLASS32 when loading a lib
pkgs/development/python-modules/ssdeep/default.nix
# can't find ar
pkgs/development/python-modules/sounddevice/default.nix
# line 3: syntax error near unexpected token `lambda'
pkgs/development/python-modules/snowflake-connector-python/default.nix
# no c compiler found
pkgs/development/python-modules/secp256k1/default.nix
#   line 1: syntax error: unexpected word (expecting ")")
pkgs/development/python-modules/pyopencl/default.nix
#/nix/store/9qsrf2ddrd1s3myyk89w94wcyw0w3idq-postgresql-13.4-lib/lib/libpq.so: file not recognized: file format not recognized
#collect2: error: ld returned 1 exit status
#Traceback (most recent call last):
#  File "/build/source/psycopg2cffi/_impl/libpq.py", line 2, in <module>
#    from psycopg2cffi._impl._libpq import ffi, lib as libpq
#ModuleNotFoundError: No module named 'psycopg2cffi._impl._libpq'
pkgs/development/python-modules/psycopg2cffi/default.nix
# some syntax error
pkgs/development/python-modules/islpy/default.nix
# no thing found setuptools
pkgs/development/python-modules/httpx/default.nix
# some syntax error
pkgs/development/python-modules/graphite-web/default.nix
# dunno the attribute name
pkgs/development/python-modules/graphite-api/default.nix
# unsupported
pkgs/development/python-modules/bwapy/default.nix
# unsupported
pkgs/development/interpreters/python/pypy/default.nix
# 2gb of deps, don't feel like testing
pkgs/applications/networking/irc/hexchat/default.nix
# An exe_wrapper is needed but was not found. Please define one in cross file and check the command and/or add it to PATH.
pkgs/applications/networking/mailreaders/notmuch/default.nix
```

~~One thing to notice is that it seems like some packages are cross-compiling, while some of their dependencies - when I cross compile them separately don't... I don't know enough python to know if this is a bug and should be a huge alarm signal or is just expected~~
~~Maybe someone else could shed some light onto this?~~
"Fixed" with https://github.com/NixOS/nixpkgs/pull/142301#issuecomment-948045461 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Artturin** [comment]
checking out the `python3Packages.cairosvg` failure

**Comment by legendofmiracles** [comment]
> I'm not sure if this fixes anything 😅
> 
> I cross compiled the bcrypt python library, and then loaded it up on my rpi - but the error still appears - just not when compiling, instead when running.
> 
> See here: https://media.discordapp.net/attachments/570351733780381697/900386597952966696/image0.jpg
> 
> I'll have to do more investigating, but it might be that all these changes achieve nothing...

Update! We can import it when we declare cffi twice! Once in `propagatedNativeBuildInputs` and once in `propagatedNativeInputs` (the latter is where it's currently)

image

I'll go ahead and add it twice in all the modified files and test the 'doesn't work' ones to see if I can fix it there

**Comment by legendofmiracles** [comment]
> Now that the dependencies failing errors also pop up, argon2_cffi is failing: `armv6l-unknown-linux-gnueabihf-gcc: error: unrecognized command-line option ‘-msse2’`

I wrote a patch to not build with that flag when cross-compiling, but now it's failing with this:
```
> extras/libargon2/src/blake2/blamka-round-opt.h:23:10: fatal error: emmintrin.h: No such file or directory
       >    23 | #include <emmintrin.h>
       >       |          ^~~~~~~~~~~~~
```
I'm pretty clueless, anyone got an idea?

**Review by Artturin — CHANGES_REQUESTED** [review]

**Comment by Artturin** [comment]
> > > > Now that the dependencies failing errors also pop up, argon2_cffi is failing: `armv6l-unknown-linux-gnueabihf-gcc: error: unrecognized command-line option ‘-msse2’`
> > > 
> > > 
> > > I wrote a patch to not build with that flag when cross-compiling, but now it's failing with this:
> > > ```
> > > > extras/libargon2/src/blake2/blamka-round-opt.h:23:10: fatal error: emmintrin.h: No such file or directory
> > >        >    23 | #include <emmintrin.h>
> > >        >       |          ^~~~~~~~~~~~~
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I'm pretty clueless, anyone got an idea?
> > 
> > 
> > this is useful for finding out what package a file is in [bennofs/nix-index](https://github.com/bennofs/nix-index)
> 
> I've tried, even after compiling `cmake` from scratch, and trying various packaging it still fails with that

i cant repro that with either `nix build ".#pkgsCross.aarch64-multiplatform.python3Packages.argon2_cffi"` or `nix build ".#pkgsCross.raspberryPi.python3Packages.argon2_cffi"`

i just get `python3.9-argon2_cffi> ModuleNotFoundError: No module named '_cffi_backend'`

**Review by Mindavi — COMMENTED** [review]
I don't feel too qualified to give a go/no go for this python related PR, but in general I guess this is good.

I did test cairosvg by cross-compiling for aarch64-multiplatform (`nix-build -A pkgsCross.aarch64-multiplatform.python39Packages.cairosvg`), running `qemu-aarch64 <path/to/aarch64/python39> ./result/bin/.cairosvg-wrapped <example svg> -o aarch64.pdf`, which did produce a valid PDF.

So it seems like this does lead to cairosvg working properly when running through an emulator, giving good confidence that this is in the right direction. For the details, I'm not sure.

**Comment by Mindavi** [comment]
Yes please, a tracking issue is always handy :)

**Comment by legendofmiracles** [comment]
I just realized an issue with the tracking issue, most of the build failures are because of non-python and non-cffi related dependencies. It's to little use, to list those, since we already know that tons and tons of packages don't cross compile.

**Comment by Mindavi** [comment]
Tracking _all_ packages that don't cross-compile is not very useful indeed, no. So maybe it's better to get this in and work futher on whatever you prefer, rather than tracking lots of packages you've no particular interest in.

**Review by jonringer — APPROVED** [review]
LGTM

```
https://github.com/NixOS/nixpkgs/pull/142301

2 packages built:
python3Packages.magic-wormhole python3Packages.paramiko
```

**Comment by lopsided98** [comment]
I was tracking down build platform contamination in the runtime closures of my cross-compiled machines, and it turns out that a lot of it is coming from this PR. Adding `cffi` to `propagatedNativeBuildInputs` causes it to become part of the runtime closure of the package (because it is referenced in `nix-support/propagated-native-build-inputs`). This is problematic because build-platform cffi pulls in a bunch of other build platform dependencies, which significantly bloat the system closure.

I previously thought this could be fixed by simply adding a `dev` output to each of the packages in this PR, which would move `nix-support/propagated-native-build-inputs` to the `dev` output. I started to do this in #200779, but I now realize that this is not enough. All downstream Python packages would also need a `dev` output as well to avoid build platform contamination.

As a concrete example, consider the `aiodns` -> `pycares` -> `cffi` dependency chain. `pycares` propagates build platform `cffi`, therefore `pycares.out` references `cffi`. If we add a `dev` output to `pycares`, `pycares.out` no longer references `cffi`, but `pycares.dev` does. `aiodns` propagates `pycares`, therefore `aiodns.out` references both `pycares.out` and `pycares.dev`, therefore build platform `cffi` is still part of the closure. If we split `aiodns`, this pushes the problem up a level, but any package that depends on `aiodns` will still have `cffi` in its closure. If every downstream package was split, this should fix the issue (since only the `out` output gets installed), but this seems infeasible.

On the other hand, does build-platform `cffi` really need to be propagated? I've only experimented with a few packages, but I was still able to build `requests` and `aiohttp` after moving `cffi` from `propagatedNativeBuildInputs` to `nativeBuildInputs` in `brotlicffi` and `pycares`.

File: pkgs/development/python-modules/liquidctl/default.nix [hunk 1]
```diff
@@ -35,6 +35,10 @@ buildPythonPackage rec {
     colorlog
   ];
 
+  propagatedNativeBuildInputs = [
+    smbus-cffi
```
Comment (lopsided98) [inline pkgs/development/python-modules/liquidctl/default.nix:39]: Also, this doesn't seem to be necessary.

**Pull request merged**